### PR TITLE
Implement PolicyV2 resource using gRPC.

### PR DESCRIPTION
**Pull request by yoursnerdly**
## Description of the change

As title.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Development

- [x] Lint rules pass locally
- [x] The code changed/added as part of this pull request has been covered with tests
- [x] All tests related to the changed code pass in development

### Code review

- [x] This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] Jira issue referenced in commit message and/or PR title

### Testing

Unit tests passed with actual CP parameters passed.

Also manually tested policy resource create, update, and delete operations.

### CODE REVIEW

File: cyral/internal/policy/v2/model.go [hunk 1]
```diff
@@ -14,127 +21,200 @@ type ChangeInfo struct {
...
+		return nil, nil
+	}
+	t, err := time.Parse(time.RFC3339, ts)
+	if err != nil {
+		return nil, fmt.Errorf("invalid %s value: %s", key, ts)
+	}
+	return timestamppb.New(t), nil
+}
+
+func timestampFromProtobuf(ts *timestamppb.Timestamp) string {
```
Comment (gengdahlCyral) [inline cyral/internal/policy/v2/model.go:106]: These timestamp helper functions I believe we will need for other migrated resources if you think it is worthwhile to collect some of these common helpers.

**Review by gengdahlCyral — APPROVED** [review]
Looks ok to me, let's also get some comments from Wilson before merging.

File: cyral/client/client.go [hunk 2]
```diff
@@ -59,15 +64,31 @@ func New(clientID, clientSecret, controlPlane string, tlsSkipVerify bool) (*Clie
 	tokenSource := tokenConfig.TokenSource(ctx)
 
 	tflog.Debug(ctx, fmt.Sprintf("TokenSource: %v", tokenSource))
+
+	grpcClient, err := grpc.NewClient(
```
[the remark this answers is not in the document]
Comment (yoursnerdly) [reply cyral/client/client.go:68]: It's the same port (443). I believe very old control plane versions had two different ports but none of them are likely in production anymore.

File: cyral/internal/policy/v2/datasource.go [hunk 3]
```diff
@@ -1,28 +1,25 @@
...
-	ResourceType:                 resourcetype.DataSource,
-	SchemaWriterFactoryGetMethod: func(_ *schema.ResourceData) core.SchemaWriter { return &PolicyV2{} },
-	ReadUpdateDeleteURLFactory: func(d *schema.ResourceData, c *client.Client) string {
-		return fmt.Sprintf("https://%s/%s/%s", c.ControlPlane, getAPIPath(d.Get("type").(string)), d.Get("id").(string))
-	},
+var dsContextHandler = core.GenericContextHandler{
+	ResourceName: dataSourceName,
+	ResourceType: resourcetype.DataSource,
+	Read:         readPolicy,
+	Create:       createPolicy,
```
[the remark this answers is not in the document]
Comment (yoursnerdly) [reply cyral/internal/policy/v2/datasource.go]: Yes, you're right. Removed.

File: cyral/internal/policy/v2/model.go (hunk 1 above)
Comment (yoursnerdly) [reply cyral/internal/policy/v2/model.go]: Done

**Review by wcmjunior — COMMENTED** [review]
My goal when I created the current `core` package was to simplify the experience for engineers adding new resources, thus it handles some things automatically, like responses for well-know operations (like a POST that will just return the id of the new element). These well-know operations are not present on the proposed handler. Any plans to add them?

File: cyral/core/generic_context_handler.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+type ResourceMethod func(context.Context, *client.Client, *schema.ResourceData) error
+
+// GenericContextHandler can be used by resource implementations to ensure that
+// the recommended best practices are consistently followed (e.g., handling of
+// 404 errors, following a create/update with a get etc). The resource implementation
+// needs to supply functions that implement the basic CRUD operations on the resource
+// using gRPC or whatever else. Note that if REST APIs are used, it is recommended
+// to use the DefaultContextHandler instead.
+type GenericContextHandler struct {
```
[the remark this answers is not in the document]
Comment (yoursnerdly) [reply cyral/core/generic_context_handler.go:27]: To be honest, I'm not particularly fond of "default" context handler. Why default?

IMO, we can rename the existing DefaultContextHandler to HTTPContextHandler or RESTContextHandler as it specifically assumes that resources are accessed using a REST API. The new handler code does not assume GRPC or any other specific protocol - hence the adjective "generic" seems appropriate to me.

What do you think @gengdahlCyral?

**Comment by yoursnerdly** [comment]
> My goal when I created the current `core` package was to simplify the experience for engineers adding new resources, thus it handles some things automatically, like responses for well-know operations (like a POST that will just return the id of the new element). These well-know operations are not present on the proposed handler. Any plans to add them?

Well the new handler does not know how the CRUD operations work (whether they read a file or do a GRPC call or whatever else), so handling "responses" in a generic/common manner is not possible (there's no notion of even a response). At most we could require the `Create` function supplied by the resource implementation to return the id which is set automatically by the `CreateHandler` method. But that does not appear to me to simplify things a whole lot and disturbs the rather pleasing symmetry that all the CRUD methods have the same signature!!

TBH, I think that the job of implementing a new resource with gRPC using the new handler object is actually pretty straightforward - the only tedious part is the conversion between protobuf objects and the resource data object - but that's required in the existing default handler as well.

Happy to discuss over a call if I'm failing to convince you here.

File: cyral/core/generic_context_handler.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (yoursnerdly) [reply cyral/core/generic_context_handler.go]: Renamed `DefaultContextHandler` to `HTTPContextHandler` and `GenericContextHandler` to just `ContextHandler`. Hope this works.

**Pull request merged**